### DISCUSSION TITLE
Welcome to rtdl Discussions!

### DISCUSSION THREAD

**Discussion by thtmnisamnstr** [realtimedatalake/rtdl#3] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by misterbykl** [comment]
Hi. I am a senior data engineer with a background of streaming and batch analytics. I am very interested in this project and would like to contribute as much as I can. Although I don't have any expertise in Go language, I would be happy to be part of this project in some way. I would be glad if you could guide me. Thanks.

**Reply by thtmnisamnstr** [reply]
That would be awesome. We’d love to find a way for you to contribute. What programming languages are you most comfortable with?

**Reply by thtmnisamnstr** [reply]
Looking at your background, it might be good for you to see if you can build a Flink Stateful Function (StateFun) to extend rtdl’s processing functionality. StateFun functions can be written in Java, Python, Go, or Javascript.

The provided `ingester` stateful function reads JSON, converts to Parquet, and writes to storage. The next one on our list is PII detection and removal/obfuscation.

If you’re interested in helping with that, let us know. Once we hear back, @dipanjanb will prep a small requirements doc explaining what we are looking for and how it should work with rtdl.

**Reply by misterbykl** [reply]
Thanks!. That would be great. I am experienced in Java, Javascript and Python. I feel more comfortable in Java. And Flink is one of my favorite framework. Let's begin with it. Looking forward for the details.

**Reply by thtmnisamnstr** [reply]
Dipanjan replied with some design guidance in the main discussion thread.

**Comment by dipanjanb** [comment]
Hi Emre,

Attaching a document where I've tried to capture my thoughts. Please go
through it and let me know in case you have any questions.

What I'd suggest is that you fork the repo, create a branch therein and
keep adding your code to that branch.

Once you're satisfied - you can raise a PR - which we can merge to the main
repo post review and subsequently merge the branch you created to the main
branch as part of a release.

Please feel free to reach out at any time should you have any questions.

Thanks & Regards,
Dipanjan

**Comment by dipanjanb** [comment]
Sorry - realized now that I replied on mail and the attachment isn't reflecting - uploaded now
[Guide to Adding Stateful Function to RTDL.docx](https://github.com/realtimedatalake/rtdl/files/8155429/Guide.to.Adding.Stateful.Function.to.RTDL.docx)

**Reply by misterbykl** [reply]
Thanks. I'll let you know once I review the doc.

**Reply by misterbykl** [reply]
@dipanjanb I reviewed the doc. It might take some time for me to understand the structure of the code. Do you have deadlines for feature developments like this?

**Reply by thtmnisamnstr** [reply]
Dioanjan replied in the main thread.
> Hi Emre,
> 
> We value community engagement.
> Hence please take your time to understand the code and let us know what
> kind of timeline you anticipate for developing the functionality outlined.
> 
> 
>

**Comment by dipanjanb** [comment]
Hi Emre,

We value community engagement.
Hence please take your time to understand the code and let us know what
kind of timeline you anticipate for developing the functionality outlined.

**Comment by misterbykl** [comment]


**Reply by misterbykl** [reply]
I tried to reproduce the issue but everything is working now. Strange. My bad I guess.

**Comment by misterbykl** [comment]
Hi Dipanjan. I've done the following so far. 
- Implemented pii-detection module in Java using state-fun structure. 
- Added the functionality of http-server, stateful-function and message type. 

I am currently working on `module.yaml `, `docker-compose.yml` and `Dockerfile`. Next step is to deploy the system and run some tests. 
Meantime, can you inform me about the pii-detection logic? Thanks!

**Comment by dipanjanb** [comment]
Hi Emre,

You can refer to this article
https://medium.com/snowflake/basic-pii-detection-and-masking-in-snowflake-using-java-1689ae63aa69
which contains sample PII detection code.

This is for creating a Java-based PII detection function for Snowflake -
you can integrate the detection logic (coded in Java) into the Stateful
Function you're building.

**Comment by misterbykl** [comment]
Hi @dipanjanb @thtmnisamnstr. Thanks again for all your help on my first PR. I can't wait for my next challenge. Let me know if there are more tasks waiting (new feature, blog post, documentation).

**Comment by dipanjanb** [comment]
You're most welcome Emre :-) Also - you did a fine job and would definitely
love to involve you in upcoming initiatives.

Thanks & Regards,
Dipanjan

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>
